revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,232 @@
+// it will be one of the errors defined in errors.go.
+//
+// (e.g. for a chain with 3 certs (leaf, intermediate, root) with both the leaf
+// and intermediate specifying 2 OCSP server, this could be a possible result):
+//
+//	[[PKIXNoCheckError, TimeoutError], [RevokedError], [nil]]
+func OCSPStatus(opts Options) [][]error {
+	certResults := make([][]error, len(opts.CertChain))
+
+	if len(opts.CertChain) == 0 {
+		return certResults
+	}
+
+	// Validate cert chain structure
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			if err := validateIssuer(cert, opts.CertChain[i+1]); err != nil {
+				fillResultsWithError(certResults, CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")})
+				return certResults
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				fillResultsWithError(certResults, CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")})
+				return certResults
+			}
+		}
+	}
+
+	// Check status for each cert in cert chain
+	var wg sync.WaitGroup
+	for i, cert := range opts.CertChain[:len(opts.CertChain)-1] {
+		wg.Add(1)
+		// Assume cert chain is accurate and next cert in chain is the issuer
+		go func(i int, cert *x509.Certificate) {
+			defer wg.Done()
+			certResults[i] = certOCSPStatus(cert, opts.CertChain[i+1], opts)
+		}(i, cert)
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	certResults[len(opts.CertChain)-1] = []error{nil}
+
+	wg.Wait()
+	return certResults
+}
+
+func fillResultsWithError(results [][]error, err error) {
+	for i := range results {
+		results[i] = []error{err}
+	}
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certOCSPStatus(cert, issuer *x509.Certificate, opts Options) []error {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return []error{NoOCSPServerError{}}
+	}
+
+	serverErrs := make([]error, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		status, err := checkStatusFromServer(cert, issuer, server, opts)
+		if err != nil {
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		// A valid response has been received from an OCSP server
+		// Result should be based on only this response, not any errors from
+		// other servers
+		switch status {
+		case ocsp.Revoked:
+			return []error{RevokedError{}}
+		case ocsp.Good:
+			return []error{nil}
+		default:
+			return []error{UnknownStatusError{}}
+		}
+
+	}
+	return serverErrs
+}
+
+func checkStatusFromServer(cert, issuer *x509.Certificate, server string, opts Options) (ocsp.ResponseStatus, error) {
```

Comment:
I feel returning two statuses (ocsp.ResponseStatus, error) from this function is causing some confusion. Why not a single error for failure and nil for successful revocation check? I think that would simplify this code a lot.

Reply:
The reason I separated these was to add more clarity into what is being returned and why. However, I can look into turning this into a single error getting returned. It just might make the process of interpreting the result of this function less understandable (since everything would be dependent on parsing the error)